homeassistant/components/image_processing/__init__.py

```diff
@@ -0,0 +1,253 @@
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    @asyncio.coroutine
+    def async_scan_service(service):
+        """Service handler for scan."""
+        image_entities = component.async_extract_from_service(service)
+
+        update_task = [entity.async_update_ha_state(True) for
+                       entity in image_entities]
+        if update_task:
+            yield from asyncio.wait(update_task, loop=hass.loop)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SCAN, async_scan_service,
+        descriptions.get(SERVICE_SCAN), schema=SERVICE_SCAN_SCHEMA)
+
+    return True
+
+
+class ImageProcessingEntity(Entity):
+    """Base entity class for image processing."""
+
+    timeout = DEFAULT_TIMEOUT  # timeout for requests
+
+    @property
+    def camera_entity(self):
+        """Return camera entity id from process pictures."""
+        return None
+
+    @property
+    def processing_class(self):
+        """Return the class of this entity from PROCESSING_CLASSES."""
+        return None
+
+    def process_image(self, image):
+        """Process image."""
+        raise NotImplementedError()
+
+    def async_process_image(self, image):
+        """Process image.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.loop.run_in_executor(None, self.process_image, image)
+
+    @property
+    def state_attributes(self):
+        """Return device specific state attributes."""
+        attr = {}
+
+        if self.processing_class is not None:
+            attr['processing_class'] = self.processing_class
+
+        return attr
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Update image and process it.
+
+        This method is a coroutine.
+        """
+        websession = async_get_clientsession(self.hass)
+        state = self.hass.states.get(self.camera_entity)
+
+        if state is None:
+            _LOGGER.warning(
+                "No entity '%s' for grab a image.", self.camera_entity)
+            return
+
+        url = "{0}{1}".format(
+            self.hass.config.api.base_url,
+            state.attributes.get(ATTR_ENTITY_PICTURE))
+
+        response = None
+        try:
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                response = yield from websession.get(url)
+
+                if response.status != 200:
+                    _LOGGER.error("Error %d on %s", response.status, url)
+                    return
+
+                image = yield from response.read()
+
+        except (asyncio.TimeoutError, aiohttp.errors.ClientError):
+            _LOGGER.error("Can't connect to %s", url)
+            return
+
+        finally:
+            if response is not None:
+                yield from response.release()
+
+        # process image data
+        yield from self.async_process_image(image)
+
+
+class ImageProcessingAlprEntity(ImageProcessingEntity):
```

Comment:
This should not live in the main component but instead in one of the two platform files.